### PR TITLE
add retry/backoff for 429s

### PR DESCRIPTION
**Pull request by alenajk**
## Status
- [x] **Ready**

## Description
Add retry and backoff for 429s

## Related PRs
- [x] This PR is independent

### CODE REVIEW

File: onecodex/api.py [hunk 1]
```diff
@@ -103,6 +105,15 @@ def __init__(
         )
         self._client._fetch_schema(cache_schema=cache_schema)
         self._session = self._client.session
+
+        # Set backoff / retry strategy for 429s
+        retry_strategy = Retry(
+            total=3, backoff_factor=4, method_whitelist=False, status_forcelist=[429]
```
Comment (jairideout) [inline onecodex/api.py:111]: We should probably add 502 and 503 codes to the list

File: onecodex/api.py [hunk 2]
```diff
@@ -103,6 +105,15 @@ def __init__(
         )
         self._client._fetch_schema(cache_schema=cache_schema)
         self._session = self._client.session
+
+        # Set backoff / retry strategy for 429s
+        retry_strategy = Retry(
```
Comment (jairideout) [inline onecodex/api.py:110]: I think the retry code in `onecodex/lib/upload.py` can be removed now that we have retries being set here

File: onecodex/api.py [hunk 3]
```diff
@@ -103,6 +105,15 @@ def __init__(
         )
         self._client._fetch_schema(cache_schema=cache_schema)
         self._session = self._client.session
+
+        # Set backoff / retry strategy for 429s
```
Comment (jairideout) [inline onecodex/api.py:109]: It'd be nice to have a unit test for the retry logic. `test_callback_retry_handling()` could be used as an example (it's testing upload retries)

Comment (jairideout) [reply onecodex/api.py:109]: Actually if we remove the retry logic from `onecodex/lib/upload.py`, the existing test is probably good enough

Comment (alenajk) [reply onecodex/api.py:108]: So, the test was using a mocked version of `session` so it was failing when I removed retry code in `onecodex/lib/upload.py` - I replaced that test with a test in `test_api_models` though

File: tests/test_api_models.py [hunk 4]
```diff
@@ -23,7 +23,17 @@ def test_api_creation(api_data):
...
+
+
+def test_retries_set_on_client_session(api_data):
+    ocx = Api(
+        api_key="1eab4217d30d42849dbde0cd1bb94e39",
+        base_url="http://localhost:3000",
+        cache_schema=False,
+    )
+
+    assert ocx._session.adapters["http://"].max_retries.total == 3
```
Comment (jairideout) [inline tests/test_api_models.py:35]: nit: it'd be nice to also test that it's set for https (that's the one we care the most about) and that `ocx._client.session` has retries set (that object path is used by the upload module & other places in the library)

**Review by jairideout — APPROVED** [review]

**Pull request merged**